File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int maxSum;

int dfs(struct TreeNode* node){
    if(node->left == NULL && node->right == NULL){
        maxSum = maxSum > node->val? maxSum: node->val;
        return node->val > 0? node->val : 0;
    }
    int left = 0, right = 0;
    if(node->left){
        left = dfs(node->left);
        left = left > 0? left: 0;
    }
    if(node->right){
        right = dfs(node->right);
        right = right > 0? right: 0;
    }
    maxSum = maxSum > node->val + left + right ? maxSum : node->val + left + right;
    return left > right? left + node->val: right + node->val;
}

int maxPathSum(struct TreeNode* root){
    maxSum = -INFINITY;
    dfs(root);
    return maxSum;
}

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        
        self.alone = float("-inf")
        def dfs(node):
            if not node.left and not node.right:
                return (node.val, float("-inf"))
            
            l, r = (float("-inf"), float("-inf")), (float("-inf"), float("-inf"))
            if node.left: l = dfs(node.left)
            if node.right: r = dfs(node.right)
                
            w, wout = max(node.val, node.val + l[0], node.val + r[0]), max(max(l), max(r))
            self.alone = max(self.alone, w, node.val + l[0] + r[0])
            
            return (w, wout)
        
        return max(max(dfs(root)), self.alone)